File: packages/sdk-python/tracelane/instrumentations/firecrawl.py
"""Firecrawl instrumentation for Tracelane.

Wraps FirecrawlApp.scrape_url and FirecrawlApp.crawl_url to emit OTel spans.
Firecrawl spans are the cost-overrun surface for document-ingestion agents
— a single unconstrained crawl can run up thousands of pages. These spans
enable scrape-quality scoring and cost attribution at the page level.

Example::

    from firecrawl import FirecrawlApp
    from tracelane.instrumentations.firecrawl import instrument_firecrawl

    app = FirecrawlApp(api_key=os.environ["FIRECRAWL_API_KEY"])
    instrument_firecrawl(app)
    result = app.scrape_url("https://example.com")
    # Span emitted with firecrawl.url and firecrawl.success
"""

from __future__ import annotations

from typing import Any

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode

_tracer = trace.get_tracer("tracelane.firecrawl", "0.1.0")


def instrument_firecrawl(app: Any) -> None:
    """Wrap a FirecrawlApp instance to emit OTel spans.

    Instruments scrape_url(), crawl_url(), and search() operations.

    Args:
        app: A firecrawl.FirecrawlApp instance.

    Note:
        URL is captured; page content is never captured.
        Pages-crawled count and success flag are recorded for cost attribution.
    """
    if hasattr(app, "scrape_url"):
        _patch_scrape_url(app)
    if hasattr(app, "crawl_url"):
        _patch_crawl_url(app)
    if hasattr(app, "search"):
        _patch_search(app)


def _patch_scrape_url(app: Any) -> None:
    original = app.scrape_url

    def _patched(url: Any = None, *args: Any, **kwargs: Any) -> Any:
        with _tracer.start_as_current_span(
            "firecrawl.scrape_url",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "firecrawl",
                "firecrawl.operation": "scrape_url",
                "firecrawl.url": str(url) if url else "",
            },
        ) as span:
            try:
                result = original(url, *args, **kwargs)
                success = getattr(result, "success", True)
                span.set_attribute("firecrawl.success", bool(success))
                content = getattr(result, "markdown", None) or getattr(result, "content", None)
                if content:
                    span.set_attribute("firecrawl.content_length", len(str(content)))
                span.set_status(
                    StatusCode.OK if success else StatusCode.ERROR,
                    "" if success else "scrape failed",
                )
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    app.scrape_url = _patched


def _patch_crawl_url(app: Any) -> None:
    original = app.crawl_url

    def _patched(url: Any = None, *args: Any, **kwargs: Any) -> Any:
        params = kwargs.get("params", args[0] if args else {})
        limit = (params or {}).get("limit", 100) if isinstance(params, dict) else 100
        with _tracer.start_as_current_span(
            "firecrawl.crawl_url",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "firecrawl",
                "firecrawl.operation": "crawl_url",
                "firecrawl.url": str(url) if url else "",
                "firecrawl.limit": int(limit),
            },
        ) as span:
            try:
                result = original(url, *args, **kwargs)
                success = getattr(result, "success", True)
                span.set_attribute("firecrawl.success", bool(success))
                data = getattr(result, "data", None)
                if isinstance(data, list):
                    span.set_attribute("firecrawl.pages_crawled", len(data))
                span.set_status(
                    StatusCode.OK if success else StatusCode.ERROR,
                    "" if success else "crawl failed",
                )
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    app.crawl_url = _patched


def _patch_search(app: Any) -> None:
    original = app.search

    def _patched(query: Any = None, *args: Any, **kwargs: Any) -> Any:
        with _tracer.start_as_current_span(
            "firecrawl.search",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "firecrawl",
                "firecrawl.operation": "search",
            },
        ) as span:
            try:
                result = original(query, *args, **kwargs)
                data = getattr(result, "data", None)
                if isinstance(data, list):
                    span.set_attribute("firecrawl.results_count", len(data))
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    app.search = _patched

File: packages/sdk-typescript/src/instrumentations/firecrawl.ts
/**
 * Firecrawl instrumentation for Tracelane.
 *
 * Wraps FirecrawlApp scrapeUrl() and crawlUrl() to emit OTel spans.
 * Firecrawl spans are the cost-overrun surface for document-ingestion agents
 * — unconstrained crawls can consume thousands of pages and dollars.
 * These spans enable cost attribution and scrape-quality scoring.
 *
 * @example
 * ```ts
 * import FirecrawlApp from "@mendable/firecrawl-js";
 * import { instrumentFirecrawl } from "@tracelanedev/sdk/firecrawl";
 *
 * const app = new FirecrawlApp({ apiKey: process.env.FIRECRAWL_API_KEY! });
 * instrumentFirecrawl(app);
 * const result = await app.scrapeUrl("https://example.com");
 * ```
 */

import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";

const tracer = trace.getTracer("@tracelanedev/sdk-firecrawl", "0.1.0");

interface FirecrawlAppLike {
	scrapeUrl: (...args: unknown[]) => Promise<unknown>;
	crawlUrl?: (...args: unknown[]) => Promise<unknown>;
	search?: (...args: unknown[]) => Promise<unknown>;
}

/**
 * Instrument a FirecrawlApp instance to emit OTel spans.
 *
 * @param app - A @mendable/firecrawl-js FirecrawlApp instance
 */
export function instrumentFirecrawl(app: FirecrawlAppLike): void {
	_patchScrapeUrl(app);
	if (app.crawlUrl) {
		_patchCrawlUrl(
			app as Required<Pick<FirecrawlAppLike, "crawlUrl">> & FirecrawlAppLike,
		);
	}
	if (app.search) {
		_patchSearch(
			app as Required<Pick<FirecrawlAppLike, "search">> & FirecrawlAppLike,
		);
	}
}

function _patchScrapeUrl(app: FirecrawlAppLike): void {
	const originalScrapeUrl = app.scrapeUrl.bind(app);

	app.scrapeUrl = async (...args: unknown[]) => {
		const url = String(args[0] ?? "");

		return tracer.startActiveSpan(
			"firecrawl.scrapeUrl",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "firecrawl",
					"firecrawl.operation": "scrapeUrl",
					"firecrawl.url": url,
				},
			},
			async (span) => {
				try {
					const result = (await originalScrapeUrl(...args)) as Record<
						string,
						unknown
					>;
					const success = Boolean(result.success ?? true);
					span.setAttribute("firecrawl.success", success);
					const content = result.markdown ?? result.content;
					if (typeof content === "string") {
						span.setAttribute("firecrawl.content_length", content.length);
					}
					span.setStatus({
						code: success ? SpanStatusCode.OK : SpanStatusCode.ERROR,
						message: success ? "" : "scrape failed",
					});
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchCrawlUrl(app: {
	crawlUrl: (...args: unknown[]) => Promise<unknown>;
}): void {
	const originalCrawlUrl = app.crawlUrl.bind(app);

	app.crawlUrl = async (...args: unknown[]) => {
		const url = String(args[0] ?? "");
		const opts = args[1] as Record<string, unknown> | undefined;
		const crawlerOpts = opts?.crawlerOptions as
			| Record<string, unknown>
			| undefined;
		const limit = Number(opts?.limit ?? crawlerOpts?.limit ?? 100);

		return tracer.startActiveSpan(
			"firecrawl.crawlUrl",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "firecrawl",
					"firecrawl.operation": "crawlUrl",
					"firecrawl.url": url,
					"firecrawl.limit": limit,
				},
			},
			async (span) => {
				try {
					const result = (await originalCrawlUrl(...args)) as Record<
						string,
						unknown
					>;
					const success = Boolean(result.success ?? true);
					span.setAttribute("firecrawl.success", success);
					const data = result.data;
					if (Array.isArray(data)) {
						span.setAttribute("firecrawl.pages_crawled", data.length);
					}
					span.setStatus({
						code: success ? SpanStatusCode.OK : SpanStatusCode.ERROR,
						message: success ? "" : "crawl failed",
					});
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchSearch(app: {
	search: (...args: unknown[]) => Promise<unknown>;
}): void {
	const originalSearch = app.search.bind(app);

	app.search = async (...args: unknown[]) => {
		return tracer.startActiveSpan(
			"firecrawl.search",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "firecrawl",
					"firecrawl.operation": "search",
				},
			},
			async (span) => {
				try {
					const result = (await originalSearch(...args)) as Record<
						string,
						unknown
					>;
					const data = result.data;
					if (Array.isArray(data)) {
						span.setAttribute("firecrawl.results_count", data.length);
					}
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}
